File: windows/tests/SSMConnect.Workflow.Tests/ConformanceTests.cs
using System.Text.Json;
using SSMConnect.Domain;
using SSMConnect.Workflow.Tests.Fixtures;

namespace SSMConnect.Workflow.Tests;

/// <summary>
/// Runs every workflow conformance fixture in <c>contracts/fixtures/workflows</c> against the real
/// workflow. This is the .NET half of AC-04: the macOS client runs the same fixtures against its
/// own implementation, and a disagreement between the two shows up here as a failing case.
/// </summary>
public sealed class ConformanceTests
{
    public static TheoryData<string> FixtureFiles()
    {
        var data = new TheoryData<string>();
        foreach (string path in FixturePaths.WorkflowFixtureFiles())
        {
            data.Add(Path.GetFileName(path));
        }

        return data;
    }

    [Fact]
    public void FixturesAreDiscovered()
    {
        IReadOnlyList<string> files = FixturePaths.WorkflowFixtureFiles();
        Assert.True(
            files.Count >= 28,
            $"Expected the contract fixtures to be copied next to the tests, found {files.Count} in " +
            $"'{FixturePaths.WorkflowsDirectory}'.");
    }

    [Theory]
    [MemberData(nameof(FixtureFiles))]
    public async Task Conforms(string fileName)
    {
        string path = Path.Combine(FixturePaths.WorkflowsDirectory, fileName);
        Fixture fixture = JsonSerializer.Deserialize<Fixture>(File.ReadAllText(path), Fixture.Options)
            ?? throw new InvalidOperationException($"Could not read fixture '{fileName}'.");

        FixtureRun run = await new FixtureRunner(fixture).RunAsync();

        AssertStates(fixture, run);
        AssertCalls(fixture, run);
        AssertForbiddenCalls(fixture, run);
        AssertTerminal(fixture, run);
    }

    private static void AssertStates(Fixture fixture, FixtureRun run)
    {
        string[] expected = fixture.Expect.States;
        string actual = string.Join(" → ", run.States);

        if (fixture.Expect.StatesAreExact)
        {
            Assert.True(
                expected.SequenceEqual(run.States),
                $"[{fixture.Id}] state sequence mismatch.\n" +
                $"  expected: {string.Join(" → ", expected)}\n" +
                $"  actual:   {actual}");
            return;
        }

        Assert.True(
            IsSubsequence(expected, run.States),
            $"[{fixture.Id}] expected states to appear in order as a subsequence.\n" +
            $"  expected: {string.Join(" → ", expected)}\n" +
            $"  actual:   {actual}");
    }

    private static bool IsSubsequence(IReadOnlyList<string> expected, IReadOnlyList<string> actual)
    {
        int cursor = 0;
        foreach (string state in actual)
        {
            if (cursor < expected.Count && expected[cursor] == state)
            {
                cursor++;
            }
        }

        return cursor == expected.Count;
    }

    /// <summary>
    /// Asserts the listed calls occurred in the listed relative order. <c>times</c> is the total
    /// count for that port and method; <c>with</c> asserts that at least one such call carried the
    /// listed arguments. Calls a fixture does not list are not constrained.
    /// </summary>
    private static void AssertCalls(Fixture fixture, FixtureRun run)
    {
        int cursor = -1;
        foreach (ExpectedCall expected in fixture.Expect.Calls)
        {
            int index = FindCall(run.Calls, expected, after: cursor);
            Assert.True(
                index >= 0,
                $"[{fixture.Id}] expected a call to {expected}{DescribeArguments(expected)} after position " +
                $"{cursor}, but the recorded sequence was:\n    {DescribeCalls(run.Calls)}");
            cursor = index;

            if (expected.Times is int times)
            {
                int actual = run.Recorder.CountOf(expected.Port, expected.Method);
                Assert.True(
                    actual == times,
                    $"[{fixture.Id}] expected {times} call(s) to {expected}, saw {actual}.\n" +
                    $"    {DescribeCalls(run.Calls)}");
            }
        }
    }

    private static void AssertForbiddenCalls(Fixture fixture, FixtureRun run)
    {
        foreach (ExpectedCall forbidden in fixture.Expect.ForbiddenCalls)
        {
            int actual = run.Recorder.CountOf(forbidden.Port, forbidden.Method);
            Assert.True(
                actual == 0,
                $"[{fixture.Id}] {forbidden} must not be called, but it was called {actual} time(s).");
        }
    }

    private static void AssertTerminal(Fixture fixture, FixtureRun run)
    {
        Terminal expected = fixture.Expect.Terminal;
        ConnectionWorkflow workflow = run.Workflow;

        Assert.True(
            workflow.State.Wire() == expected.State,
            $"[{fixture.Id}] terminal state was '{workflow.State.Wire()}', expected '{expected.State}'. " +
            $"States seen: {string.Join(" → ", run.States)}");

        if (expected.ErrorCategory is string category)
        {
            Assert.True(
                workflow.ErrorCategory.Wire() == category,
                $"[{fixture.Id}] error category was '{workflow.ErrorCategory.Wire()}', expected '{category}'. " +
                $"Message: {workflow.ErrorMessage ?? "<none>"}");
        }

        if (expected.ErrorMessageContains is string fragment)
        {
            Assert.True(
                workflow.ErrorMessage?.Contains(fragment, StringComparison.OrdinalIgnoreCase) == true,
                $"[{fixture.Id}] expected the error message to contain '{fragment}', got " +
                $"'{workflow.ErrorMessage ?? "<none>"}'.");
        }

        if (expected.WarningPresent is bool warning)
        {
            Assert.True(
                (workflow.WarningMessage is not null) == warning,
                $"[{fixture.Id}] expected warningPresent={warning}, warning was " +
                $"'{workflow.WarningMessage ?? "<none>"}'.");
        }

        if (expected.InstanceId is not null || ExpectsNullInstanceId(fixture))
        {
            Assert.True(
                workflow.InstanceId == expected.InstanceId,
                $"[{fixture.Id}] instance ID was '{workflow.InstanceId ?? "<null>"}', expected " +
                $"'{expected.InstanceId ?? "<null>"}'.");
        }

        if (expected.TunnelActive is bool tunnelActive)
        {
            Assert.True(
                workflow.TunnelActive == tunnelActive,
                $"[{fixture.Id}] expected tunnelActive={tunnelActive}, was {workflow.TunnelActive}.");
        }

        if (expected.PasswordInMemory is bool password)
        {
            Assert.True(
                (workflow.Password is not null) == password,
                $"[{fixture.Id}] expected passwordInMemory={password}, was {workflow.Password is not null}.");
        }
    }

    /// <summary>
    /// A fixture that explicitly writes <c>"instanceId": null</c> is asserting the field was
    /// cleared, which is different from not mentioning it at all.
    /// </summary>
    private static bool ExpectsNullInstanceId(Fixture fixture)
    {
        string raw = File.ReadAllText(Path.Combine(FixturePaths.WorkflowsDirectory, $"{fixture.Id}.json"));
        using JsonDocument document = JsonDocument.Parse(raw);
        return document.RootElement.TryGetProperty("expect", out JsonElement expect)
            && expect.TryGetProperty("terminal", out JsonElement terminal)
            && terminal.TryGetProperty("instanceId", out JsonElement instanceId)
            && instanceId.ValueKind == JsonValueKind.Null;
    }

    private static int FindCall(IReadOnlyList<RecordedCall> calls, ExpectedCall expected, int after)
    {
        for (int index = after + 1; index < calls.Count; index++)
        {
            RecordedCall call = calls[index];
            if (call.Port == expected.Port && call.Method == expected.Method && Matches(call, expected))
            {
                return index;
            }
        }

        return -1;
    }

    private static bool Matches(RecordedCall call, ExpectedCall expected)
    {
        if (expected.With is null)
        {
            return true;
        }

        foreach ((string name, JsonElement value) in expected.With)
        {
            if (!call.Arguments.TryGetValue(name, out object? actual))
            {
                return false;
            }

            string expectedText = value.ValueKind == JsonValueKind.String ? value.GetString() ?? "" : value.ToString();
            if (!string.Equals(Convert.ToString(actual, System.Globalization.CultureInfo.InvariantCulture),
                    expectedText, StringComparison.Ordinal))
            {
                return false;
            }
        }

        return true;
    }

    private static string DescribeArguments(ExpectedCall expected) =>
        expected.With is null
            ? string.Empty
            : " with " + string.Join(", ", expected.With.Select(pair => $"{pair.Key}={pair.Value}"));

    private static string DescribeCalls(IReadOnlyList<RecordedCall> calls) =>
        calls.Count == 0 ? "<no calls>" : string.Join("\n    ", calls.Select((call, i) => $"{i}: {call}"));
}

File: SSMConnectKit/Tests/SSMConnectKitTests/Conformance/ConformanceTests.swift
import Foundation
import Testing
@testable import SSMConnectKit

/// Runs every workflow conformance fixture in `contracts/fixtures/workflows` against the real
/// `ConnectionStateMachine`. This is the macOS half of AC-04: the Windows client runs the same
/// fixtures against `SSMConnect.Workflow`, and a disagreement between the two clients shows up
/// here, or there, as a failing case.
@Suite("Workflow conformance (contracts/fixtures/workflows)")
struct ConformanceTests {
    static let fixtureFiles: [URL] = FixturePaths.workflowFixtureFiles()

    @Test("every contract fixture is discovered")
    func fixturesAreDiscovered() {
        #expect(
            Self.fixtureFiles.count >= 28,
            "Expected at least 28 workflow fixtures in \(FixturePaths.workflowsDirectory.path), found \(Self.fixtureFiles.count)."
        )
    }

    @Test("conforms", arguments: fixtureFiles)
    @MainActor
    func conforms(file: URL) async throws {
        let fixture = try FixturePaths.loadFixture(at: file)
        let run = try await FixtureRunner(fixture: fixture).run()

        assertStates(fixture, run)
        assertCalls(fixture, run)
        assertForbiddenCalls(fixture, run)
        assertTerminal(fixture, run)
    }

    // MARK: - States

    private func assertStates(_ fixture: Fixture, _ run: FixtureRun) {
        let expected = fixture.expect.states
        let actual = run.states
        let rendering = """
            [\(fixture.id)] state sequence mismatch.
              expected: \(expected.joined(separator: " → "))
              actual:   \(actual.joined(separator: " → "))
            """

        if fixture.expect.statesAreExact {
            #expect(expected == actual, "\(rendering)")
        } else {
            #expect(isSubsequence(expected, of: actual), "\(rendering)  (expected as a subsequence)")
        }
    }

    private func isSubsequence(_ expected: [String], of actual: [String]) -> Bool {
        var cursor = 0
        for state in actual where cursor < expected.count && expected[cursor] == state {
            cursor += 1
        }
        return cursor == expected.count
    }

    // MARK: - Calls

    /// Asserts the listed calls occurred in the listed relative order. `times` is the total count
    /// for that port and method; `with` requires at least one such call to carry those arguments.
    /// Calls a fixture does not list are not constrained.
    private func assertCalls(_ fixture: Fixture, _ run: FixtureRun) {
        var cursor = -1
        for expected in fixture.expect.calls {
            guard let index = findCall(run.calls, expected, after: cursor) else {
                Issue.record("""
                    [\(fixture.id)] expected a call to \(expected)\(describe(expected.with)) after \
                    position \(cursor), but the recorded sequence was:
                        \(describe(run.calls))
                    """)
                continue
            }
            cursor = index

            if let times = expected.times {
                let actual = run.recorder.count(port: expected.port, method: expected.method)
                #expect(
                    actual == times,
                    """
                    [\(fixture.id)] expected \(times) call(s) to \(expected), saw \(actual).
                        \(describe(run.calls))
                    """
                )
            }
        }
    }

    private func assertForbiddenCalls(_ fixture: Fixture, _ run: FixtureRun) {
        for forbidden in fixture.expect.forbiddenCalls {
            let actual = run.recorder.count(port: forbidden.port, method: forbidden.method)
            #expect(actual == 0, "[\(fixture.id)] \(forbidden) must not be called, but it was called \(actual) time(s).")
        }
    }

    private func findCall(_ calls: [RecordedCall], _ expected: ExpectedCall, after: Int) -> Int? {
        calls.indices.dropFirst(after + 1).first { index in
            let call = calls[index]
            return call.port == expected.port && call.method == expected.method && matches(call, expected)
        }
    }

    private func matches(_ call: RecordedCall, _ expected: ExpectedCall) -> Bool {
        guard let with = expected.with else { return true }
        return with.allSatisfy { name, value in
            call.arguments[name]?.comparableText == value.comparableText
        }
    }

    // MARK: - Terminal

    @MainActor
    private func assertTerminal(_ fixture: Fixture, _ run: FixtureRun) {
        let expected = fixture.expect.terminal
        let machine = run.machine
        let trail = "States seen: \(run.states.joined(separator: " → "))"

        #expect(
            machine.state.rawValue == expected.state,
            "[\(fixture.id)] terminal state was '\(machine.state.rawValue)', expected '\(expected.state)'. \(trail)"
        )

        if let category = expected.errorCategory {
            #expect(
                machine.errorCategory.rawValue == category,
                """
                [\(fixture.id)] error category was '\(machine.errorCategory.rawValue)', \
                expected '\(category)'. Message: \(machine.errorMessage ?? "<none>")
                """
            )
        }

        if let fragment = expected.errorMessageContains {
            #expect(
                machine.errorMessage?.localizedCaseInsensitiveContains(fragment) == true,
                "[\(fixture.id)] expected the error message to contain '\(fragment)', got '\(machine.errorMessage ?? "<none>")'."
            )
        }

        if let warning = expected.warningPresent {
            #expect(
                (machine.warningMessage != nil) == warning,
                "[\(fixture.id)] expected warningPresent=\(warning), warning was '\(machine.warningMessage ?? "<none>")'."
            )
        }

        if let instanceId = expected.instanceId {
            #expect(
                machine.instanceId == instanceId,
                "[\(fixture.id)] instance ID was '\(machine.instanceId ?? "<null>")', expected '\(instanceId ?? "<null>")'."
            )
        }

        if let localPort = expected.localPort {
            #expect(
                machine.localPort == localPort,
                "[\(fixture.id)] local port was \(machine.localPort.map(String.init) ?? "<null>"), expected \(localPort.map(String.init) ?? "<null>")."
            )
        }

        if let tunnelActive = expected.tunnelActive {
            #expect(
                machine.tunnelActive == tunnelActive,
                "[\(fixture.id)] expected tunnelActive=\(tunnelActive), was \(machine.tunnelActive)."
            )
        }

        if let passwordInMemory = expected.passwordInMemory {
            #expect(
                (machine.password != nil) == passwordInMemory,
                "[\(fixture.id)] expected passwordInMemory=\(passwordInMemory), was \(machine.password != nil)."
            )
        }
    }

    // MARK: - Rendering

    private func describe(_ with: [String: JSONValue]?) -> String {
        guard let with, !with.isEmpty else { return "" }
        return " with " + with.keys.sorted().map { "\($0)=\(with[$0]!.comparableText ?? "null")" }.joined(separator: ", ")
    }

    private func describe(_ calls: [RecordedCall]) -> String {
        calls.isEmpty
            ? "<no calls>"
            : calls.enumerated().map { "\($0.offset): \($0.element)" }.joined(separator: "\n        ")
    }
}
